flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElection.java

```diff
@@ -39,17 +44,37 @@ public interface LeaderElection extends AutoCloseable {
      *
      * @param leaderSessionID The new leader session ID
      * @param leaderAddress The address of the new leader
+     * @return A future that completes successfully if the confirmation succeeded or fails
+     *     exceptionally with a {@link LeadershipLostException} if the leadership was revoked in the
+     *     meantime.
+     */
+    default CompletableFuture<Void> confirmLeadershipAsLeader(
+            UUID leaderSessionID, String leaderAddress) {
+        checkNotNull(leaderSessionID);
+        checkNotNull(leaderAddress);
+
+        return runAsLeader(
+                leaderSessionID, () -> confirmLeadership(leaderSessionID, leaderAddress));
+    }
+
+    /**
+     * Runs the actual leadership confirmation. This method should only be called if the leadership
+     * is acquired. Otherwise, a {@link IllegalStateException} should be thrown.
      */
     void confirmLeadership(UUID leaderSessionID, String leaderAddress);
 
     /**
-     * Returns {@code true} if the service's {@link LeaderContender} has the leadership under the
-     * given leader session ID acquired.
+     * Runs the given {@code callback} on the leader election's main thread only if the associated
+     * {@code leaderSessionId} is still valid.
      *
-     * @param leaderSessionId identifying the current leader
-     * @return true if the associated {@link LeaderContender} is the leader, otherwise false
+     * @param leaderSessionId The session ID that's associated with the given {@code callback}.
+     * @param callback The callback that shall be executed as a leader.
+     * @return The future referring to the result of the callback operation. This future would
+     *     complete exceptionally with a {@link LeadershipLostException} if the leadership wasn't
+     *     active anymore.
      */
-    boolean hasLeadership(UUID leaderSessionId);
+    CompletableFuture<Void> runAsLeader(
```

Comment:
yeeeeah I'm not sold on this.

I struggle to understand the contract as to what should be run via the method.

Is it that everything called in the LeaderContender callbacks must used this method?
It feels very wrong that this method is used to access the JRS for example (even if that internally does things asynchronously), when that has little to do with leader _election_, which runs counter to the idea of running as few things as possible in the leader election executor.

I also don't quite follow yet why such a big API change is necessary.
Couldn't we have saved ourselves a lot of work by just changing hasLeadership to return a `CompletableFuture<Boolean>`?
Now we have this strange API where a sync confirmLeadership method exists but must basically never be called from one side because it risks dead-locks.

Can't we solve the lock acquisition problem by removing the need to call hasLeadership?
We already have an API that signals whether we are the leader or not; can't we write that outside of the contender lock into an AtomicBoolean and use that instead (while ensuring that grant/revoke don't do any heavy lifting)?